### ISSUE TITLE
GraphQL transform directive proper usage of @connection vs @key

### ISSUE THREAD

**Issue by nateiler** [aws-amplify/amplify-cli#1881]
** Which Category is your question related to? **
API

** What AWS Services are you utilizing? **
Amplify / AppSync / DynamoDb

** Provide additional details e.g. code snippets **
Rather new to Amplify; stubbing out some concepts which I could use a nudge when it comes to best practices.  The example I'm working on is Tenant -> Contact model structure.  

The requirement are rather straight forward:
- Allow filtering of Contacts based on Tenant (tenantId)

The Tenant model
```
type Tenant 
  @model
{
  id: ID!
  name: String!
  createdAt: String
}
```

---

The Contact model (v1)
```
type Contact 
  @model
{
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail
  tenant: Tenant! @connection
  createdAt: String
}
```
Using the `@connection` directive the relationships is established rather easily. However, it appears the 'Tenant' object is attached to every 'Contact' which I don't necessarily need.  The major downside of this is I can't seem to get all Contacts by Tenant.

---

The Contact model (v2)
```
type Contact 
  @model
{
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail
  tenant: Tenant! @connection(keyField: "tenantId")
  tenantId: ID!
  createdAt: String
}
```
I noticed the 'keyField' usage referenced in other GH issues and using it allows filtering by 'tenantId' which satisfies my requirement.   However, Contacts are created, but I'm not able to fetch any via query returned.  When creating, the following error is returned (although the record is added in Dynamo).  

```
{
	"data": {
		"listContacts": {
			"items": [
				null
			]
		}
	},
	"errors": [{
		"locations": null,
		"message": "Cannot return null for non-nullable type: 'Tenant' within parent 'Contact' (/listContacts/items[0]/tenant)"
	}]
}
```

---

The Contact model (v3)
```
type Company 
  @model
  @key(name: "ByTenant", fields: ["tenantId", "createdAt"], queryField: "contactsByTenant")
{
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail
  tenantId: ID!
  createdAt: String
}
```
Not using the `@connection` directive and adding a @key.  This is the closest and most successful representation of what I'm looking for.  My concern is around not using `@connection` and if it's something I should re-consider. 

---

In summary, my take away is (correct me if I'm wrong):
@connection is a simple way to establish/link a bi-directional index(s)
@connection automatically attaches the related model on the object
@connection doesn't allow filtering by the related model's primary key
@key is a more verbose way to establish an index
@key doesn't attach a related model (because one isn't defined)

I come from a relational db background; so it's taking a bit to adapt.

**Comment by mikeparisstuff** [comment]
The **@connection** directive was made available before the **@key** directive and does two things 1) creates a GSI 2) configures a resolver that relates the objects. The **@key** directive is a more general approach to solving part (1) and we have a work item to build a new directive that solves part (2). Stay tuned for an RFC covering the design.

In the meantime, if you want to make **@connection**-like resolvers between types you can

1) Use `@connection` as it not being removed or deprecated. Make sure you get your `keyField` and `sortField` correct.
2) Use `@key` to define your indexes and use [custom resolvers](https://aws-amplify.github.io/docs/cli-toolchain/graphql#custom-resolvers) to interact with the indexes. As soon as the new directive is available you will be able to replace your custom resolvers.

You can also move from **@connection** to **@key** in a backwards compatible way so you can start with **@connection** and move to **@key**.

**Comment by nateiler** [comment]
Thanks!

Very helpful.  It took a bit of prototyping, but I've managed to construct a schema that is working well.

I was stumbling with how to set the `keyField` value on a @connection and managed to figure it out.  Ref: https://github.com/aws-amplify/amplify-cli/issues/1570#issuecomment-519599845
